fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStoreTest.java

```diff
@@ -1336,4 +1352,163 @@ public void recordTypesWithOneOfUnion() {
+        }
+    }
+
+    @EnumSource(TestHelpers.BooleanEnum.class)
+    @ParameterizedTest(name = "noUnionImplicitUsage [repeatSaveOrDoUpdate = {0}]")
+    public void noUnionImplicitUsage(TestHelpers.BooleanEnum repeatSaveOrDoUpdate) {
+        int version;
+
+        // MyOtherRecord has no explicit usage. The proto file has a union and does not include MyOtherRecord.
+        try (FDBRecordContext context = fdb.openContext()) {
+            openMetaDataStore(context);
+            metaDataStore.saveRecordMetaData(TestRecordsImplicitUsageProto.getDescriptor());
+            context.commit();
+            version = metaDataStore.getRecordMetaData().getVersion();
+            assertNotNull(metaDataStore.getRecordMetaData().getRecordType("MySimpleRecord"));
+            MetaDataException e = assertThrows(MetaDataException.class, () -> metaDataStore.getRecordMetaData().getRecordType("MyOtherRecord"));
+            assertEquals(e.getMessage(), "Unknown record type MyOtherRecord");
+        }
+
+        // The evolved proto no longer has a union. The record types with no explicit usage should now automatically show up in the union.
+        try (FDBRecordContext context = fdb.openContext()) {
+            openMetaDataStore(context);
+            assertEquals(version, metaDataStore.getRecordMetaData().getVersion());
+            if (repeatSaveOrDoUpdate.toBoolean()) {
+                metaDataStore.updateRecords(TestRecordsImplicitUsageNoUnionProto.getDescriptor());
+            } else {
+                metaDataStore.saveRecordMetaData(TestRecordsImplicitUsageNoUnionProto.getDescriptor());
+            }
+            assertNotNull(metaDataStore.getRecordMetaData().getRecordType("MySimpleRecord"));
+            assertNotNull(metaDataStore.getRecordMetaData().getRecordType("MyOtherRecord"));
+            assertEquals(version + 1, metaDataStore.getRecordMetaData().getVersion());
+            context.commit();
+        }
+    }
+
+
+    @Test
+    public void noUnionLocalFileDescriptor() {
+        try (FDBRecordContext context = fdb.openContext()) {
+            openMetaDataStore(context);
+            metaDataStore.saveRecordMetaData(TestNoUnionProto.getDescriptor());
+            context.commit();
+        }
+
+        // The type cannot become NESTED in the local file descriptor
+        try (FDBRecordContext context = fdb.openContext()) {
+            openMetaDataStore(context);
+            metaDataStore.setLocalFileDescriptor(TestNoUnionEvolvedIllegalProto.getDescriptor());
+            MetaDataException e = assertThrows(MetaDataException.class, () -> metaDataStore.getRecordMetaData());
+            assertEquals("record type removed from union", e.getMessage());
+            context.commit();
+        }
+
+        // Deprecate a record type.
+        try (FDBRecordContext context = fdb.openContext()) {
+            openMetaDataStore(context);
+            deprecateRecordType("com.apple.foundationdb.record.testnounion.MySimpleRecord");
```

Comment:
Does this only work with fully-qualified names like this?

Looking ahead, it will be relatively easy to support checking the type name matches for record types when specified fully-fully (e.g., `.com.apple.foundationdb.record.testnounion.MySimpleRecord`) and fully unqualified (e.g., `MySimpleRecord`), but not middlingly qualified (e.g., `foundationdb.record.testnounion.MySimpleRecord` or actually `com.apple.foundationdb.record.testnounion.MySimpleRecord` without the leading `.`). I'm concerned about the API breaking in a kind of subtle way here in that certain strings specifying type names might stop working.

The other thing that occurs to me is that for types in the union, their least-qualified name (e.g., `MySimpleRecord`) needs to be unique, so a possible contract could be that you only need to give it that value. I think, by necessity, this won't match how `renameRecordType`'s contract needs to look (I think).

Reply:
Yeah valid concern.

We currently use the `fullName` of the type for setting the type name of the union fields in union-less local files. Let me prepend "." to it, so it's at least in harmony with `addDefaultUnionIfMissing`. 

`deprecateRecordType` works with the type name. When it comes from the proto, it's fully qualified (with "."). However, `addRecordType` currently uses the record type name, which I agree that it has to be  consistent. This is unrelated to this change though. I'll file an issue for this and fix it soon.